### PR TITLE
Change HandleClient to process correctly the data from the TCP socket.

### LINKED ISSUE #4272: Receiving large amounts of data on syslog over TCP causes wrong output for analysis.
|Wazuh version|Component|Install type|
|---|---|---|
| 3.12 | remoted | Manager |

When receiving a large amount of logs on a TCP syslog connection the remoted component send an incorrect output for analysis. There are basically two different conditions/errors.
- When several log messages are received on the same TCP packet, only the first log message is correctly parsed and passed down for analysis.
- When several log messages are received in 2 or more TCP packets, only the first log message is correctly parsed and passed down for analysis and, upon receiving the second message, the remainder of the log messages from the first packet are logged with the first log of the second packet. The remainder of the log messages from the second packet are discarded.

Both of this errors are because the logic on the HandleClient function assumes that the received TCP packet is either a full log message, part of the message with the rest of it coming in following messages or a full log message and a part of a second message. If the received TCP packet holds two full log messages, two full messages and the beginning of a third, three messages and the beginning of a fourth, etc., the message passed down for analysis will be wrong.

To proof this, I wrote a simple python program that sends the message `<1000>Dec 25 20:45:02 MyHost example[12345]: User 'admin' logged from '192.168.1.100'\n` 800 times on the syslog port of a manager and the archives log showed the following:

```
{"timestamp":"2019-11-22T22:55:13.999+0100","rule":{"level":4,"description":"User logged","id":"100010","firedtimes":13,"mail":false,"groups":["local","syslog","sshd"]},"agent":{"id":"000","name":"manager"},"manager":{"name":"manager"},"id":"1574459713.2405631","full_log":"Dec 25 20:45:02 MyHost example[12345]: User 'admin' logged from '192.168.1.100'","predecoder":{"program_name":"example","timestamp":"Dec 25 20:45:02","hostname":"MyHost"},"decoder":{"name":"example"},"data":{"srcip":"192.168.1.100","dstuser":"admin"},"location":"192.168.50.10"}
{"timestamp":"2019-11-22T22:57:09.200+0100","rule":{"level":4,"description":"User logged","id":"100010","firedtimes":14,"mail":false,"groups":["local","syslog","sshd"]},"agent":{"id":"000","name":"manager"},"manager":{"name":"manager"},"id":"1574459829.2414018","full_log":"Dec 25 20:45:02 MyHost example[12345]: User 'admin' logged from '192.168.1.100'\n<1000>Dec 25 20:45:02 MyHost example[12345]: User 'admin' logged from '192.168.1.100'\n<1000>Dec 25 20:45:02 MyHost example[12345]: User 'admin' logged from '192.168.1.100'\n<1000>Dec 25 20:45:02 MyHost example[12345]: User 'admin' logged from '192.168.1.100'\n<1000>Dec 25 20:45:02 MyHost example[12345]: User 'admin' logged from '192.168.1.100'\n<1000>Dec 25 20:45:02 MyHost example[12345]: User 'admin...
```
The easiest solution I can see to this problem is to check whether the received TCP packet contains anymore full log messages before the cleanup part of the following code and send all of them in separate messages, maybe wrapping around the logic for parsing syslog messages until the buffer length is 0 or no other line feeds are found.
https://github.com/wazuh/wazuh/blob/5b02b003d0b57f2b262e1bd0ca32524eaee30e39/src/remoted/syslogtcp.c#L99-L137

### PR DESCRIPTION
**Pull request by antoniomanuelfr**
|Related issue|
|---|
|#4272|

## Description
This PR aims to fix #4272 by changing the function that gets the data from the TCP buffer and send the messages correctly to syslog. 

## Tests 
In order to test that the implementation is correct, I have written the next python script: 
```python
import socket
import os
import time

msg = "<1000>Dec 25 20:45:02 MyHost example[{}]: User 'admin' logged from '192.168.1.100'\n"
server = socket.socket( socket.AF_INET, socket.SOCK_STREAM )
server.connect(("10.2.0.10", 1514))
for i in range(0,800):
    sent = server.send(msg.format(i).encode())

server.close()
```
## Configuration options
- Activate the logall option in order to get all the information that arrives at the manager.
- Activate TCP socket in syslog: 
```
  <remote>
    <connection>syslog</connection>
    <port>1514</port>
    <protocol>tcp</protocol>
    <allowed-ips>10.2.0.0/24</allowed-ips>
  </remote>
```

## Logs/Alerts example
```
2020 May 22 11:53:02 centosmanager->10.2.0.1 <1000>Dec 25 20:45:02 MyHost example[793]: User 'admin' logged from '192.168.1.100'
2020 May 22 11:53:02 centosmanager->10.2.0.1 <1000>Dec 25 20:45:02 MyHost example[794]: User 'admin' logged from '192.168.1.100'
2020 May 22 11:53:02 centosmanager->10.2.0.1 <1000>Dec 25 20:45:02 MyHost example[795]: User 'admin' logged from '192.168.1.100'
2020 May 22 11:53:02 centosmanager->10.2.0.1 <1000>Dec 25 20:45:02 MyHost example[796]: User 'admin' logged from '192.168.1.100'
2020 May 22 11:53:02 centosmanager->10.2.0.1 <1000>Dec 25 20:45:02 MyHost example[797]: User 'admin' logged from '192.168.1.100'
2020 May 22 11:53:02 centosmanager->10.2.0.1 <1000>Dec 25 20:45:02 MyHost example[798]: User 'admin' logged from '192.168.1.100'
2020 May 22 11:53:02 centosmanager->10.2.0.1 <1000>Dec 25 20:45:02 MyHost example[799]: User 'admin' logged from '192.168.1.100'
```

- Compilation without warnings in every supported platform
  - [x] Linux
- [x] Source installation
- [x] Source upgrade

- Memory tests for Linux
  - [x] Scan-build report
  - [x] Coverity
  - [x] Valgrind (memcheck and descriptor leaks check)

### CODE REVIEW

**Comment by wazuhci** [comment]
## Analysis_dynamic_valgrind logcollector analysis:
**Commit:** 42a54773d00991fcecc449a097f2d9dd7d2b654a
**Parameters:** track-origins="yes" leak-check="full" keep-stacktraces="alloc-and-free" track-fds="yes" num-callers="20" show-leak-kinds="definite,indirect"
**Agent config:** [ossec.conf](https://ci.wazuh.com/data/Analysis_dynamic_valgrind/v3.13.0/B2088/ossec.conf)
**Agent logs:** [ossec.log](https://ci.wazuh.com/data/Analysis_dynamic_valgrind/v3.13.0/B2088/ossec.log)
**Coverage:** 

```json
Lines executed:3.93% of 1859

```
**Coverage report detail:** [coverage-report-summary](https://ci.wazuh.com/data/Analysis_dynamic_valgrind/v3.13.0/B2088/coverage-report-detail)
**Valgrind result:** [Analysis_dynamic_valgrind-v3.13.0-B2088.log](https://ci.wazuh.com/data/Analysis_dynamic_valgrind/v3.13.0/B2088/Analysis_dynamic_valgrind-v3.13.0-B2088.log)

**Comment by antoniomanuelfr** [comment]
**Scan-build:** No related errors.
**Valgrind:** No related errors.

**Comment by wazuhci** [comment]
## Analysis_dynamic_valgrind logcollector analysis:
**Commit:** f948088e114c1e69ae8d9394cd225295b44116b8
**Parameters:** track-origins="yes" leak-check="full" keep-stacktraces="alloc-and-free" track-fds="yes" num-callers="20" show-leak-kinds="definite,indirect"
**Agent config:** [ossec.conf](https://ci.wazuh.com/data/Analysis_dynamic_valgrind/v3.13.0/B2202/ossec.conf)
**Agent logs:** [ossec.log](https://ci.wazuh.com/data/Analysis_dynamic_valgrind/v3.13.0/B2202/ossec.log)
**Coverage:** 

```json
Lines executed:31.41% of 1859

```
**Coverage report detail:** [coverage-report-summary](https://ci.wazuh.com/data/Analysis_dynamic_valgrind/v3.13.0/B2202/coverage-report-detail)
**Valgrind result:** [Analysis_dynamic_valgrind-v3.13.0-B2202.log](https://ci.wazuh.com/data/Analysis_dynamic_valgrind/v3.13.0/B2202/Analysis_dynamic_valgrind-v3.13.0-B2202.log)

**Comment by wazuhci** [comment]
## Analysis_dynamic_valgrind syscheckd analysis:
**Commit:** f948088e114c1e69ae8d9394cd225295b44116b8
**Parameters:** track-origins="yes" leak-check="full" keep-stacktraces="alloc-and-free" track-fds="yes" num-callers="20" show-leak-kinds="definite,indirect"
**Agent config:** [ossec.conf](https://ci.wazuh.com/data/Analysis_dynamic_valgrind/v3.13.0/B2203/ossec.conf)
**Agent logs:** [ossec.log](https://ci.wazuh.com/data/Analysis_dynamic_valgrind/v3.13.0/B2203/ossec.log)
**Coverage:** 

```json
Lines executed:26.90% of 3190

```
**Coverage report detail:** [coverage-report-summary](https://ci.wazuh.com/data/Analysis_dynamic_valgrind/v3.13.0/B2203/coverage-report-detail)
**Valgrind result:** [Analysis_dynamic_valgrind-v3.13.0-B2203.log](https://ci.wazuh.com/data/Analysis_dynamic_valgrind/v3.13.0/B2203/Analysis_dynamic_valgrind-v3.13.0-B2203.log)

File: src/remoted/syslogtcp.c [hunk 1]
```diff
@@ -50,98 +73,23 @@ static void HandleClient(int client_socket, char *srcip)
...
     /* Initialize some variables */
     memset(buffer, '\0', OS_MAXSTR + 2);
-    memset(storage_buffer, '\0', OS_MAXSTR + 2);
-    memset(tmp_buffer, '\0', OS_MAXSTR + 2);
-
-
     while (1) {
         /* If an error occurred, or received 0 bytes, we need to return and close the socket */
-        r_sz = OS_RecvTCPBuffer(client_socket, buffer, OS_MAXSTR - 2);
+        r_sz = recv(client_socket, buffer + buff_size, (OS_MAXSTR - buff_size) - 2, 0);
```
Comment (vikman90) [inline src/remoted/syslogtcp.c:78]: - `buffer` is `OS_MAXSTR + 2` bytes. Why do we read up to `OS_MAXSTR - 2`?
- `recv` does not null-terminate the buffer, as `OS_RecvTCPBuffer` does.

Should we use this?
```c
r_sz = recv(client_socket, buffer + buff_size, OS_MAXSTR - buff_size);
```

File: src/remoted/syslogtcp.c [hunk 2]
```diff
@@ -31,17 +31,40 @@ static int OS_IPNotAllowed(char *srcip)
...
 
+/**
+ * @brief Function that sends a buffer to a queue.
+ * @param buffer String with the received contents from the TCP socket.
+ * @param buff_size Size of the buffer.
+ * @param srcip String with the IP of the queue where the message will be sent.
+ */
+void send_buffer(char *buffer, int *buff_size, char *srcip) {
+    int offset;
+    char *buffer_pt = strchr(buffer, '\n');
```
Comment (vikman90) [inline src/remoted/syslogtcp.c:42]: `strchr` Assumes that the currently valid part of `buffer` is null-terminated.

File: src/remoted/syslogtcp.c [hunk 3]
```diff
@@ -31,17 +31,40 @@ static int OS_IPNotAllowed(char *srcip)
...
+
+            if ((logr.m_queue = StartMQ(DEFAULTQUEUE, WRITE)) < 0) {
+                merror_exit(QUEUE_FATAL, DEFAULTQUEUE);
+            }
+        }
+        // Re-calculate the used size of buffer and remove the message from the buffer
+        *buff_size = *buff_size - (offset + 1);
+        memcpy(buffer, buffer_pt + 1, *buff_size);
+        // Find the next '\n'
+        buffer_pt = strchr(buffer, '\n');
```
Comment (vikman90) [inline src/remoted/syslogtcp.c:60]: `strchr` assumes that `memcpy` copied the buffer including the ending `0`.

**Comment by wazuhci** [comment]
## Analysis_dynamic_valgrind logcollector analysis:
**Commit:** 7f638ccdb179057281cbf5aafe72eb4a9eaa8aae
**Parameters:** track-origins="yes" leak-check="full" keep-stacktraces="alloc-and-free" track-fds="yes" num-callers="20" show-leak-kinds="definite,indirect"
**Agent config:** [ossec.conf](https://ci.wazuh.com/data/Analysis_dynamic_valgrind/v3.13.0/B2393/ossec.conf)
**Agent logs:** [ossec.log](https://ci.wazuh.com/data/Analysis_dynamic_valgrind/v3.13.0/B2393/ossec.log)
**Coverage:** 

```json
Lines executed:31.41% of 1859

```
**Coverage report detail:** [coverage-report-summary](https://ci.wazuh.com/data/Analysis_dynamic_valgrind/v3.13.0/B2393/coverage-report-detail)
**Valgrind result:** [Analysis_dynamic_valgrind-v3.13.0-B2393.log](https://ci.wazuh.com/data/Analysis_dynamic_valgrind/v3.13.0/B2393/Analysis_dynamic_valgrind-v3.13.0-B2393.log)

**Comment by wazuhci** [comment]
## Analysis_dynamic_valgrind syscheckd analysis:
**Commit:** 7f638ccdb179057281cbf5aafe72eb4a9eaa8aae
**Parameters:** track-origins="yes" leak-check="full" keep-stacktraces="alloc-and-free" track-fds="yes" num-callers="20" show-leak-kinds="definite,indirect"
**Agent config:** [ossec.conf](https://ci.wazuh.com/data/Analysis_dynamic_valgrind/v3.13.0/B2394/ossec.conf)
**Agent logs:** [ossec.log](https://ci.wazuh.com/data/Analysis_dynamic_valgrind/v3.13.0/B2394/ossec.log)
**Coverage:** 

```json
Lines executed:26.90% of 3190

```
**Coverage report detail:** [coverage-report-summary](https://ci.wazuh.com/data/Analysis_dynamic_valgrind/v3.13.0/B2394/coverage-report-detail)
**Valgrind result:** [Analysis_dynamic_valgrind-v3.13.0-B2394.log](https://ci.wazuh.com/data/Analysis_dynamic_valgrind/v3.13.0/B2394/Analysis_dynamic_valgrind-v3.13.0-B2394.log)

**Comment by wazuhci** [comment]
## Analysis_dynamic_valgrind logcollector analysis:
**Commit:** 482fae9ee7922efce8506256fddfc85a9691a680
**Parameters:** track-origins="yes" leak-check="full" keep-stacktraces="alloc-and-free" track-fds="yes" num-callers="20" show-leak-kinds="definite,indirect"
**Agent config:** [ossec.conf](https://ci.wazuh.com/data/Analysis_dynamic_valgrind/v3.13.0/B2423/ossec.conf)
**Agent logs:** [ossec.log](https://ci.wazuh.com/data/Analysis_dynamic_valgrind/v3.13.0/B2423/ossec.log)
**Coverage:** 

```json
Lines executed:31.41% of 1859

```
**Coverage report detail:** [coverage-report-summary](https://ci.wazuh.com/data/Analysis_dynamic_valgrind/v3.13.0/B2423/coverage-report-detail)
**Valgrind result:** [Analysis_dynamic_valgrind-v3.13.0-B2423.log](https://ci.wazuh.com/data/Analysis_dynamic_valgrind/v3.13.0/B2423/Analysis_dynamic_valgrind-v3.13.0-B2423.log)

**Comment by wazuhci** [comment]
## Analysis_dynamic_valgrind syscheckd analysis:
**Commit:** 482fae9ee7922efce8506256fddfc85a9691a680
**Parameters:** track-origins="yes" leak-check="full" keep-stacktraces="alloc-and-free" track-fds="yes" num-callers="20" show-leak-kinds="definite,indirect"
**Agent config:** [ossec.conf](https://ci.wazuh.com/data/Analysis_dynamic_valgrind/v3.13.0/B2424/ossec.conf)
**Agent logs:** [ossec.log](https://ci.wazuh.com/data/Analysis_dynamic_valgrind/v3.13.0/B2424/ossec.log)
**Coverage:** 

```json
Lines executed:26.90% of 3190

```
**Coverage report detail:** [coverage-report-summary](https://ci.wazuh.com/data/Analysis_dynamic_valgrind/v3.13.0/B2424/coverage-report-detail)
**Valgrind result:** [Analysis_dynamic_valgrind-v3.13.0-B2424.log](https://ci.wazuh.com/data/Analysis_dynamic_valgrind/v3.13.0/B2424/Analysis_dynamic_valgrind-v3.13.0-B2424.log)

**Pull request merged**